### PR TITLE
Qute docs bugs

### PR DESCRIPTION
**Pull request by mkouba**

### CODE REVIEW

**Review by cescoffier — APPROVED** [review]

File: docs/src/main/asciidoc/qute-reference.adoc [hunk 1]
```diff
@@ -513,6 +513,28 @@ This section might also come in handy when we'd like to avoid multiple expensive
 ----
 <1> `it` is the result of `item.callExpensiveLogicToGetTheValue(1,'foo',bazinga)`. The method is only invoked once even though the result may be used in multiple expressions.
 
+===== Let/Set Section
+
+This section is an alternative to `{#with}` which allows to specify multiple named parameters:
```
Comment (FroMage) [inline docs/src/main/asciidoc/qute-reference.adoc:518]: We should still remove `#{with expr as binding}` because it does not have the semantics of `with`, and can be done with `#{let binding = expr}`.

File: docs/src/main/asciidoc/qute-reference.adoc [hunk 2]
```diff
@@ -567,30 +589,30 @@ Let's suppose we have a template called `item.html`:
...
 We can include the tag like this:
 
 [source,html]
 ----
 <ul>
-{#each items}
+{#for item in items}
   <li>
-  {#item this showImage=true /} <1>
+  {#myItem item showImage=true /} <1>
```
Comment (FroMage) [inline docs/src/main/asciidoc/qute-reference.adoc:610]: Huh, why did you rename it? Does the `#item` tag namespace clash with the local variable `item`? If that's the case it's a bug.

Comment (FroMage) [inline docs/src/main/asciidoc/qute-reference.adoc:610]: This is the first place I realised how we can pass more than one parameter to a tag. Is this formally documented somewhere else in this file? If not, it should, I was sure I would need a comma.

File: independent-projects/qute/core/src/test/java/io/quarkus/qute/UserTagTest.java [hunk 3]
```diff
@@ -17,17 +17,17 @@ public void testUserTag() {
...
-        assertEquals("Herbert", template1.render(Collections.singletonMap("order", order)));
-        assertEquals("nope", template2.render(Collections.singletonMap("order", order)));
-        assertEquals("nope", template3.render(Collections.singletonMap("order", order)));
-        assertEquals("nope", template4.render(Collections.singletonMap("order", order)));
+        assertEquals("Herbert",
+                engine.parse("{#myTag order showImage=true /}").render(Collections.singletonMap("order", order)));
+        assertEquals("nope", engine.parse("{#myTag order /}").render(Collections.singletonMap("order", order)));
+        assertEquals("nope", engine.parse("{#myTag showImage=false /}").render(Collections.singletonMap("order", order)));
+        assertEquals("nope",
+                engine.parse("{#each this}{#myTag showImage=false /}{/each}").render(Collections.singletonMap("order", order)));
```
Comment (FroMage) [inline independent-projects/qute/core/src/test/java/io/quarkus/qute/UserTagTest.java:28]: As I see tests mentioning `this` I have to ask what it refers to?

**Review by FroMage — CHANGES_REQUESTED** [review]
I added some remarks/questions.

File: docs/src/main/asciidoc/qute-reference.adoc (hunk 2 above)
Comment (mkouba) [reply docs/src/main/asciidoc/qute-reference.adoc:610]: No, it could be `item` or anything else. I just wanted to have the example more readable... `{#item item}` is not good.

Comment (FroMage) [reply docs/src/main/asciidoc/qute-reference.adoc:610]: OK, pfew :)

But frankly `#{item item}` is still my better than `#{myItem item}`. Perhaps it's `#{itemDescription item}` you were aiming for? ;)

Comment (mkouba) [reply docs/src/main/asciidoc/qute-reference.adoc:610]: Comma was never required. It's not documented. The whole SPI needs docs but since it's still a WIP it has a lower priority. In short, a section/tag may consume any number of named/unnamed parameters and it's the responsibility of a section to interprete the params, i.e. it could be an expression ('item.name') or it could be ignored (such as `in` in the `{#for item in items}`).

We already have https://github.com/quarkusio/quarkus/issues/7326.

File: independent-projects/qute/core/src/test/java/io/quarkus/qute/UserTagTest.java (hunk 3 above)
Comment (mkouba) [reply independent-projects/qute/core/src/test/java/io/quarkus/qute/UserTagTest.java:28]: In this particular case it refers to the data object passed to the template. `this` always returns the last context object on the stack.

File: docs/src/main/asciidoc/qute-reference.adoc (hunk 1 above)
Comment (mkouba) [reply docs/src/main/asciidoc/qute-reference.adoc:518]: Ok. The semantics of `with` is not well-defined in Qute (or in templates in general) but I agree that it could be replaced with let/set.

File: docs/src/main/asciidoc/qute-reference.adoc (hunk 2 above)
Comment (mkouba) [reply docs/src/main/asciidoc/qute-reference.adoc:610]: > But frankly `#{item item}` is still way better than `#{myItem item}`.

Oh, really? ;-) Anyway, `#{itemDescription item}` or `{#itemDetail item}` reads definitely better. My intention was to show the link between the tag name registered above. But this is not important in quarkus where tags are registered automatically.

File: docs/src/main/asciidoc/qute-reference.adoc (hunk 1 above)
Comment (FroMage) [reply docs/src/main/asciidoc/qute-reference.adoc:518]: Well, `with` to modify the current scope is fine, it's the version with `as` which is problematic.

Comment (mkouba) [reply docs/src/main/asciidoc/qute-reference.adoc]: Done.

File: docs/src/main/asciidoc/qute-reference.adoc (hunk 2 above)
Comment (FroMage) [reply docs/src/main/asciidoc/qute-reference.adoc]: So, this is what bugs me about the regularity of those tags. Some use `in` or `as` keywords (not evaluated) while others use `=` which seems to be an operator, but I'm not clear on that either.

The semantics are not clear to me as a user. And it doesn't appear regular.

Compare this with the following syntax:

```
<ul>
#{list items:products, as:'product'}
    <li>${product}</li>
#{/list}
</ul>
```

Where you can pass exactly one unnamed parameter to a tag, or you pass them by name with the `:` operator and `,` separator. As you can see, here `products` is an expression, while `'product'` is a string. The semantics are clear.

It's less flexible, but it's much more consistent with programming language constructs. For example, I'm not sure how the current semantics of passing parameters as-is to Java tag implementation combines with the rule for named parameters: can I call `#{foo first second third=something}` and expect it to work if `foo` is a Java tag?

Using the current Qute syntax, this would mean:

```
{#for items, as='item'}
```

Now, given that we only have `as` in the problematic `with` form (which we should remove in favour of `let`/`set`), that only leaves us with a single keyword `in`, and I've observed people mostly write custom tags using taglet files (not in Java), perhaps there's not much risk in this sort of keybord behaviour proliferating to confuse users.

But at the very least, we need to document that parameters are passed separated by spaces, named parameters use `=`.

File: independent-projects/qute/core/src/test/java/io/quarkus/qute/UserTagTest.java (hunk 3 above)
Comment (FroMage) [reply independent-projects/qute/core/src/test/java/io/quarkus/qute/UserTagTest.java:28]: If it's not documented, it should not be implemented. If it's implemented it should be documented.
ATM I don't see the value of having `this` be implemented and resolved, so I suggest we disable it.

File: docs/src/main/asciidoc/qute-reference.adoc [hunk 4]
```diff
@@ -489,29 +489,40 @@ This could be useful to simplify the template structure:
...
+  {/if}
 {/with}
 ----
-<1> `myParent` is the alias that can be used inside the tag.
+<1> `this` is the result of `item.callExpensiveLogicToGetTheValue(1,'foo',bazinga)`. The method is only invoked once even though the result may be used in multiple expressions.
 
-This section might also come in handy when we'd like to avoid multiple expensive invocations:
+===== Let/Set Section
 
+This section is an alternative to `{#with}` which allows to specify one or more named parameters:
```
Comment (FroMage) [inline docs/src/main/asciidoc/qute-reference.adoc:508]: No it is not: it does not change the scope: it adds new lexical bindings.

```suggestion
This section allows you to define named local variables:
```

File: docs/src/main/asciidoc/qute-reference.adoc [hunk 5]
```diff
@@ -489,29 +489,40 @@ This could be useful to simplify the template structure:
...
 <2> The `description` will be also resolved against the `item.parent`.
 
-It's also possible to specify an alias for the context object:
+This section might also come in handy when we'd like to avoid multiple expensive invocations:
 
 [source,html]
 ----
-{#with item.parent as myParent} <1>
-  <h1>{myParent.name}</h1>
+{#with item.callExpensiveLogicToGetTheValue(1,'foo',bazinga)}
```
Comment (FroMage) [inline docs/src/main/asciidoc/qute-reference.adoc:496]: Damnit, `this` is back. But at least this time it makes sense.

File: docs/src/main/asciidoc/qute-reference.adoc [hunk 6]
```diff
@@ -489,29 +489,40 @@ This could be useful to simplify the template structure:
...
-{#with item.callExpensiveLogicToGetTheValue(true) as it}
-  {#if it is "fun"} <1>
-    <h1>Yay!</h1>
-  {#else}
-    <h1>{it} is not fun at all!</h1>
-  {/if}
-{/with}
+{#let myParent=order.item.parent}
+  <h1>{myParent.name}</h1>
+  <p>Price: {price}
```
Comment (FroMage) [inline docs/src/main/asciidoc/qute-reference.adoc:513]: This `price` thing confuses me: it's not coming from `order.item.parent/myParent`, right?

**Review by FroMage — CHANGES_REQUESTED** [review]
That's much better. I still have a few questions.

File: docs/src/main/asciidoc/qute-reference.adoc (hunk 4 above)
Comment (mkouba) [reply docs/src/main/asciidoc/qute-reference.adoc:508]: > No it is not: it does not change the scope: it adds new lexical bindings.

Not really. Under the hood it does modify the current context. But I can change the wording. No problem.

File: docs/src/main/asciidoc/qute-reference.adoc (hunk 6 above)
Comment (mkouba) [reply docs/src/main/asciidoc/qute-reference.adoc:513]: Nope. It's a mistake. Good catch! The original version had a `price` param too. Normally, Qute would attempt to find a `price` in the parent contexts...

Comment (FroMage) [reply docs/src/main/asciidoc/qute-reference.adoc:513]: Well, it's fine that it would resolve it from the parent scope (that's what I expect), I just wanted to be sure, because its presence in this example confused me.

File: docs/src/main/asciidoc/qute-reference.adoc (hunk 4 above)
Comment (FroMage) [reply docs/src/main/asciidoc/qute-reference.adoc]: OK, fine, it does introduce a new scope, right? It doesn't modify the current scope in the sense that it modifies the existing scope, right?

I still think that in this context, saying something along the lines of "it defines new local variables" is the easiest thing to explain to devs, no? We just want to make it clear that it's a bounded local scope with new bindings.

**Review by FroMage — APPROVED** [review]
Perfect, thanks!
